CMakeLists.txt

```diff
@@ -58,15 +58,17 @@ option(LLAMA_SANITIZE_ADDRESS           "llama: enable address sanitizer"
 option(LLAMA_SANITIZE_UNDEFINED         "llama: enable undefined sanitizer"                     OFF)
 
 # instruction set specific
-option(LLAMA_AVX                        "llama: enable AVX"                                     ON)
-option(LLAMA_AVX2                       "llama: enable AVX2"                                    ON)
-option(LLAMA_AVX512                     "llama: enable AVX512"                                  OFF)
-option(LLAMA_AVX512_VBMI                "llama: enable AVX512-VBMI"                             OFF)
-option(LLAMA_AVX512_VNNI                "llama: enable AVX512-VNNI"                             OFF)
-option(LLAMA_FMA                        "llama: enable FMA"                                     ON)
-# in MSVC F16C is implied with AVX2/AVX512
-if (NOT MSVC)
-    option(LLAMA_F16C                   "llama: enable F16C"                                    ON)
+if (NOT LLAMA_NATIVE)
+    option(LLAMA_AVX                         "llama: enable AVX"                                ON)
+    option(LLAMA_AVX2                        "llama: enable AVX2"                               ON)
+    option(LLAMA_AVX512                      "llama: enable AVX512"                             OFF)
+    option(LLAMA_AVX512_VBMI                 "llama: enable AVX512-VBMI"                        OFF)
+    option(LLAMA_AVX512_VNNI                 "llama: enable AVX512-VNNI"                        OFF)
+    option(LLAMA_FMA                         "llama: enable FMA"                                ON)
+    # in MSVC F16C is implied with AVX2/AVX512
+    if (NOT MSVC)
+        option(LLAMA_F16C                    "llama: enable F16C"                               ON)
+    endif()
```

Comment:
They still exist if LLAMA_NATIVE is enabled, it's just that the default is not set and it's intepreted as false. So if you compile with `-DLLAMA_AVX512=ON -DLLAMA_NATIVE=ON` you'll get `-march=native -mavx512f -mavx512bw`.

The alternate option generates the same result and still requires an if/else (cmake doesn't support inverting an option variable).
```
# instruction set specific
if (LLAMA_NATIVE)
    set(INS_ENB OFF)
else()
    set(INS_ENB ON)
endif()

option(LLAMA_AVX                        "llama: enable AVX"                                     ${INS_ENB})
option(LLAMA_AVX2                       "llama: enable AVX2"                                    ${INS_ENB})
option(LLAMA_AVX512                     "llama: enable AVX512"                                  OFF)
option(LLAMA_AVX512_VBMI                "llama: enable AVX512-VBMI"                             OFF)
option(LLAMA_AVX512_VNNI                "llama: enable AVX512-VNNI"                             OFF)
option(LLAMA_FMA                        "llama: enable FMA"                                     ${INS_ENB})
# in MSVC F16C is implied with AVX2/AVX512
if (NOT MSVC)
    option(LLAMA_F16C                   "llama: enable F16C"                                    ${INS_ENB})
endif()
```

Reply:
The options are not set, and they also have no description. So if someone wanted to use the CMake GUI, they would not see the options in the list and would have to manually add them. Right?

Reply:
There is `cmake_dependent_option` which I think could solve this: https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html#module:CMakeDependentOption

Reply:
> The options are not set, and they also have no description. So if someone wanted to use the CMake GUI, they would not see the options in the list and would have to manually add them. Right?

I didn't even know CMake had a gui :eyes: but it loads the options fine with descriptions even when they're behind the if statement.
![screenshot](https://github.com/ggerganov/llama.cpp/assets/139727413/86f242ee-6aa7-4ebc-a011-e7952ba5f2c2)

> There is `cmake_dependent_option` which I think could solve this: https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html#module:CMakeDependentOption

If I use that then it'll hide the options when LLAMA_NATIVE is set, which is not what we want right?

Anyways I've switched to the IS_ENB version in my latest commit.

Reply:
> it loads the options fine with descriptions even when they're behind the if statement.

True, but only because LLAMA_NATIVE currently defaults to false. If it defaulted to true, the CPU feature options would not appear until you disable it. Then they would be present until you clear the cache, even if you re-enabled LLAMA_NATIVE.

Reply:
the options should disappear in the gui if you delete the `CMakeCache.txt` and run cmake with `-DLLAMA_NATIVE=ON`.
@Cebtenzzre is right.